tmail-backend/webadmin/webadmin-rate-limit/src/main/scala/com/linagora/tmail/webadmin/RateLimitPlanManagementRoutes.scala

```diff
@@ -0,0 +1,163 @@
+import com.linagora.tmail.webadmin.model.{GetAllRateLimitPlanResponseDTO, OperationLimitationsDTO, RateLimitationDTO, RateLimitingPlanCreateRequestDTO, RateLimitingPlanDTO, RateLimitingPlanResetRequestDTO}
+import javax.inject.Inject
+import org.apache.james.rate.limiter.api.AllowedQuantity
+import org.apache.james.webadmin.Constants.SEPARATOR
+import org.apache.james.webadmin.Routes
+import org.apache.james.webadmin.utils.{ErrorResponder, JsonExtractor, JsonTransformer, Responses}
+import org.eclipse.jetty.http.HttpStatus.NOT_FOUND_404
+import reactor.core.scala.publisher.{SFlux, SMono}
+import spark.{Request, Response, Route, Service}
+
+import scala.jdk.CollectionConverters._
+import scala.jdk.OptionConverters._
+import scala.jdk.StreamConverters._
+
+class RateLimitPlanManagementRoutes @Inject()(planRepository: RateLimitationPlanRepository,
+                                              jsonTransformer: JsonTransformer) extends Routes {
+  private val PLAN_ID_PARAM = ":planId"
+  private val PLAN_NAME_PARAM = ":planName"
+  private val BASE_PATH =  s"${SEPARATOR}rate-limit-plans"
+  private val CREATE_A_PLAN_PATH = s"$BASE_PATH$SEPARATOR$PLAN_NAME_PARAM"
+  private val UPDATE_A_PLAN_PATH = s"$BASE_PATH$SEPARATOR$PLAN_ID_PARAM"
+  private val GET_A_PLAN_PATH = s"$BASE_PATH$SEPARATOR$PLAN_ID_PARAM"
+  private val GET_ALL_PLAN_PATH = BASE_PATH
+
+  val createRequestExtractor = new JsonExtractor[RateLimitingPlanCreateRequestDTO](classOf[RateLimitingPlanCreateRequestDTO])
+  val resetRequestExtractor = new JsonExtractor[RateLimitingPlanResetRequestDTO](classOf[RateLimitingPlanResetRequestDTO])
+
+  override def getBasePath: String = BASE_PATH
+
+  override def define(service: Service): Unit = {
+    service.post(CREATE_A_PLAN_PATH, createAPlan, jsonTransformer)
+    service.put(UPDATE_A_PLAN_PATH, updateAPlan, jsonTransformer)
+    service.get(GET_A_PLAN_PATH, getAPlan, jsonTransformer)
+    service.get(GET_ALL_PLAN_PATH, getAllPlan, jsonTransformer)
+  }
+
+  private def createAPlan: Route = (request: Request, response: Response) =>
+    SMono.fromPublisher(planRepository.create(toCreateRequest(request)))
+      .map(toRateLimitingPlanDTO)
+      .block()
+
+  private def updateAPlan: Route = (request: Request, response: Response) =>
+    SMono.fromPublisher(planRepository.update(toResetRequest(request)))
+      .onErrorResume { case e: RateLimitingPlanNotFoundException => SMono.error(
+        ErrorResponder.builder()
+        .statusCode(NOT_FOUND_404)
+        .`type`(ErrorResponder.ErrorType.NOT_FOUND)
+        .message("Plan does not exist")
+        .haltError())
+      }
+      .`then`(SMono.just(Responses.returnNoContent(response)))
+      .block()
+
+  private def getAPlan: Route = (request: Request, response: Response) =>
+    SMono.fromPublisher(planRepository.get(extractRateLimitingPlanId(request)))
+      .map(toRateLimitingPlanDTO)
+      .onErrorResume { case e: RateLimitingPlanNotFoundException => SMono.error(
```

Comment:
Should we use
`service.exception(...)` to handler exception, then we don't need "handle"` multi-time. 

```java
        service.exception(IllegalArgumentException.class, (ex, req, res) -> {
            LOGGER.info("Invalid arguments supplied in the user request", ex);
            res.status(BAD_REQUEST_400);
            res.body(ErrorResponder.builder()
                .statusCode(BAD_REQUEST_400)
                .type(INVALID_ARGUMENT)
                .message("Invalid arguments supplied in the user request")
                .cause(ex)
                .asString());
        });
```

Reply:
IMO the RateLimitingPlanNotFoundException use cases in this Route and other Routes (if we gonna have them) may not be generic and duplicated enough so I propose to keep it like this. Also not every route gonna throw this exception, explicitly handling this can emphasize the logic here.

Reply:
Webadmin already handle IllegalArgumentException centrally

Reply:
> IMO the RateLimitingPlanNotFoundException use cases in this Route and other Routes (if we gonna have them) may not be generic and duplicated enough so I propose to keep it like this. Also not every route gonna throw this exception, explicitly handling this can emphasize the logic here

yes, I believe we can check `url path` with a regex pattern to handler it for special routes we want. 
(That is the way Spring do), 
of course, it is one more view, not necessary to change. 
The current code looks good